core/shared/src/main/scala/scalaz/zio/stream/Stream.scala

```diff
@@ -564,20 +564,31 @@ trait Stream[-R, +E, +A] extends Serializable { self =>
             q2: Queue[Take[E2, B]],
             s: S
           ): ZIO[R2, E2, S] = {
-            val takeLeft  = if (leftDone) IO.succeed(None) else Take.option(q1.take)
-            val takeRight = if (rightDone) IO.succeed(None) else Take.option(q2.take)
-
-            takeLeft.zip(takeRight).flatMap {
-              case (left, right) =>
-                f(left, right) match {
-                  case None => IO.succeed(s)
-                  case Some(c) =>
-                    g(s, c).flatMap { s =>
-                      if (cont(s)) loop(left.isEmpty, right.isEmpty, q1, q2, s)
-                      else IO.succeed(s)
-                    }
-                }
-            }
+            val takeLeft: ZIO[R2, E2, Option[A]]  = if (leftDone) IO.succeed(None) else Take.option(q1.take)
```

Comment:
I had to annotate these lines explictly, otherwise this is inferred as `ZIO[Any, E2, Option[A]]` and line 584 results in an error that `Any` is required but `R2` is found.

I haven't investigated why the narrowing fails, but this might be another candidate for inference improvements before 1.0.